text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+| `NeglectStabilty`   | Stability   | `transmute_{from,into}`, `unsafe_transmute_{from,into}` |
+| `NeglectAlignment`  | Safety      | `unsafe_transmute_{from,into}`                          |
+| `NeglectValidity`   | Soundness   | `unsafe_transmute_{from,into}`                          |
+
+`NeglectStabilty` implements the `SafeTransmuteOptions` and `TransmuteOptions` marker traits, as it can be used in both safe and unsafe code. The selection of multiple options is encoded by grouping them as a tuple; e.g., `(NeglectAlignment, NeglectValidity)` is a selection of both the `NeglectAlignment` and `NeglectValidity` options.
+
+We introduce two marker traits which serve to group together the options that may be used with safe transmutes, and those which may be used with `unsafe` transmutes:
+```rust
+pub trait SafeTransmuteOptions: private::Sealed
+{}
+
+pub trait TransmuteOptions: SafeTransmuteOptions
+{}
+
+impl SafeTransmuteOptions for () {}
+impl TransmuteOptions for () {}
+```
+
+#### `NeglectStability`
+[`NeglectStability`]: #neglectstability
+
+By default, `TransmuteFrom` and `TransmuteInto`'s methods require that the [layouts of the source and destination types are SemVer-stable][stability]. The `NeglectStability` option disables this requirement.
+```rust
+pub struct NeglectStability;
+
+impl SafeTransmuteOptions for NeglectStability {}
+impl TransmuteOptions for NeglectStability {}
+```
+
+Prior to the adoption of the [stability declaration traits][stability], crate authors documented the layout guarantees of their types with doc comments. The `TransmuteFrom` and `TransmuteInto` traits and methods may be used with these types by requesting that the stability check is neglected; for instance:
+
+```rust
+fn serialize<W: Write>(val : LibraryType, dst: W) -> std::io::Result<()>
+where
+    LibraryType: TransmuteInto<[u8; size_of::<LibraryType>()], NeglectStability>
+{
+    ...
+}
+```
+
+Neglecting stability over-eagerly cannot cause unsoundness or unsafety. For this reason, it is the only transmutation option available on the safe methods `transmute_from` and `transmute_into`. However, neglecting stability over-eagerly may cause your code to cease compiling if the authors of the source and destination types make changes that affect their layout.
+
+By using the `NeglectStability` option to transmute types you do not own, you are committing to ensure that your reliance on these types' layouts is consistent with their documented stability guarantees.
+
+#### `NeglectAlignment`
+[ext-ref-casting]: #NeglectAlignment
+
+By default, `TransmuteFrom` and `TransmuteInto`'s methods require that, when transmuting references, the minimum alignment of the destination's referent type is no greater than the minimum alignment of the source's referent type. The `NeglectAlignment` option disables this requirement.
+```rust
+pub struct NeglectAlignment;
+
+impl TransmuteOptions for NeglectAlignment {}
+```
```

Comment:
The docs here should be very careful to say that it is *not ever* okay to neglect alignment without a dynamic check. Creating unaligned references is insta-UB (as y'all know but having this option available could make it seem like that's something unsafe code might legitimately do).

Reply:
Will do! I'll tackle this in my next round of edits.

Just to confirm: it *is* okay to construct a mis-aligned raw pointer, just not to dereference it. In other words, for references you must do the check *before* conversion; for raw pointers, it's okay to do it *after*?

Reply:
> Just to confirm: it is okay to construct a mis-aligned raw pointer, just not to dereference it. In other words, for references you must do the check before conversion; for raw pointers, it's okay to do it after?

Yes, it is safe to construct misaligned raw pointer. [`1 as *const u32` is valid safe rust code today](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=59986f4e332d1b33651a3b861eaee59e), and similarly pointer casts between types of different alignment is also safe. The alignment requirement must be upheld at deref time.